### ISSUE TITLE
Multi Threading discord.py

### ISSUE THREAD

**Issue by kooldamian28** [Rapptz/discord.py#7380]
### Summary

Multi-Threading discord.py may be necessary as the bot grows and has more users to ensure fast performance and stability.

### What is the feature request for?

discord.ext.commands

### The Problem

Multi-Threading the discord.py library may be necessary as the bot grows, this can be seen in `Carl-bot`, which was previously made in Python but was remade in Rust, however, I do not know if the owner of the bot moved from Python to Rust for the reason stated above, I have seen a lot of bots made in Rust that are multi-threaded. I mainly think not multi-threading might be a bottleneck as the bot gains more users. I have researched on making my bot multi-threaded, but I found that [most asyncio objects are not thread safe.](https://python.readthedocs.io/en/stable/library/asyncio-dev.html#concurrency-and-multithreading) So multi-threading bot commands manually is unsafe, which is why I think multi-threading the library is better.

### The Ideal Solution

The best solution would be to multi-thread the library probably using [`threading`](https://docs.python.org/3/library/threading.html) or [`multiprocessing`](https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing) to fully leverage threads on a machine/server. Or another, but less preferred solution would be to make the asyncio objects thread-safe.

**Comment by vicky5124** [comment]
CarlBot main reason to switch away from discord.py is cache scalability, it's rewrite is being done with the Twilight library rather than the more common Serenity, as it allows him to do what he needs.
![15 27 20_13-08-2021](https://user-images.githubusercontent.com/8344197/129364185-d6e2533c-2ddc-4a45-be62-efbe133afdf8.png)
CarlBot is not the only one suffering from this, to a lesser extent, GearBot is on the same boat, rewriting it's bot to Rust's twilight for much better scaling on cache and performance, in this case, mainly to be able to use Redis for the cache.

He has solved this issue on d.py by using a new instance of the bot for each shard, AFAIK no multithreading is used, this allows the cache to be more manageable, even if it's still massive, and even throughput doesn't slow down because of python being single threaded.

On another note, `threading` runs only on the current interpreter, which is single threaded, it works more like concurrency and doesn't take advantage of multi-core parallelism. `multiprocessing` on the other hand, spawns other interpreters that can execute multiple things in different cores, CPUs or even machines, but transferring data between the interpreters is still single threaded, managed by the main thread, which is slow as it's limited by the GIL (Global Interpreter Lock).

**Comment by ChillFish8** [comment]
The feature you want can already be done, you just require manual sharding, the way you manage your processes/clusters is up to you and IMO should stay that way, generally, Kubernetes stateful sets work well or you'll want something like a master-slave worker system where workers are able to get some info on what shards they're in control of.
 
If this was an inbuilt thing with Discord.py I feel like it would add a lot of complexity with process management and handling which in reality can be done better when left to tools designed for the job.

TL;DR -> Python Discord bots scale in the same sense webservers scale, if you want to use more than one CPU and be able to grow to handle more events.
